### PR TITLE
fix: deflake run_kubectl_apply_tests

### LINKED ISSUE #93651: [Flaky Test] run_kubectl_apply_tests (ci-kubernetes-integration-master)
**Which jobs are flaking**:

`integration master (ci-kubernetes-integration-master)`

**Which test(s) are flaking**:

`test-cmd.run_kubectl_apply_tests `

**Testgrid link**:

https://testgrid.k8s.io/sig-release-master-blocking#integration-master

**Reason for failure**:

```
Error from server (NotFound): resources.mygroup.example.com "myobj" not found
```

**Anything else we need to know**:

/sig cli
/cc @kubernetes/ci-signal @soltysh 
/milestone v1.20
/prority important-soon

### PR DESCRIPTION
**Pull request by knight42**
Signed-off-by: knight42 <anonymousknight96@gmail.com>

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature

/kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #93651 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/cmd/apply.sh [hunk 1]
```diff
@@ -444,6 +444,9 @@ run_kubectl_server_side_apply_tests() {
 }
 __EOF__
 
+  # Ensure the CRD has been created
```
Comment (liggitt) [inline test/cmd/apply.sh:447]: This is not the correct check... the CRD will always be created here

The issue is whether the API server has recognized and started serving the associated CR API yet

Comment (seans3) [reply test/cmd/apply.sh]: @liggitt Is it possible to do a wait on a Condition for a CRD? (e.g. `kubectl wait --for=condition=Ready mygroup.example.com/myobj --timeout=30s`)

Comment (liggitt) [reply test/cmd/apply.sh]: I think we want to wait until kubectl observes the expected API available in discovery (`kubectl api-resources` contains the new API)

Comment (hasheddan) [reply test/cmd/apply.sh]: @seans3 you could wait for condition `Established`, although I am not certain that guarantees that it is being served https://github.com/kubernetes/kubernetes/blob/2585d78c79054a6d85620eb5430a1498bb747967/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types.go#L279

File: test/cmd/apply.sh [hunk 2]
```diff
@@ -444,6 +444,16 @@ run_kubectl_server_side_apply_tests() {
...
 __EOF__
 
+  # Ensure the API server has recognized and started serving the associated CR API
+  local tries=5
+  for i in $(seq 1 $tries); do
+      if kubectl "${kube_flags_with_token[@]}" api-resources --api-group 'mygroup.example.com' -oname | grep -q 'resources.mygroup.example.com'; then
+          break
+      fi
+      echo "${i}: Waiting for CR API to be available"
+      sleep $((i-1))
```
Comment (hasheddan) [inline test/cmd/apply.sh:454]: I may be missing something but why the changing sleep time? Is it intended to be a rudimentary backoff?

Comment (knight42) [reply test/cmd/apply.sh:454]: no special reason, just follow the existing practice https://github.com/kubernetes/kubernetes/blob/c3816157f97cc068600b0dbebdbe80733ee5c3eb/hack/lib/test.sh#L82-L93

Comment (knight42) [reply test/cmd/apply.sh:454]: we could fix the sleep time if that is preferred

Comment (hasheddan) [reply test/cmd/apply.sh:454]: @knight42 I think it's okay, just wanted to know the motivation :)

**Comment by hasheddan** [comment]
Saw [another flake](https://prow.k8s.io/view/gcs/kubernetes-jenkins/logs/ci-kubernetes-integration-master/1290888190839754752) on this last night. @liggitt could you give this another look?

File: test/cmd/apply.sh [hunk 3]
```diff
@@ -444,6 +444,16 @@ run_kubectl_server_side_apply_tests() {
 }
 __EOF__
 
+  # Ensure the API server has recognized and started serving the associated CR API
+  local tries=5
+  for i in $(seq 1 $tries); do
+      if kubectl "${kube_flags_with_token[@]}" api-resources --api-group 'mygroup.example.com' -oname | grep -q 'resources.mygroup.example.com'; then
```
Comment (liggitt) [inline test/cmd/apply.sh:450]: piping to `grep -q` breaks things, see https://github.com/kubernetes/kubernetes/pull/57805

File: test/cmd/apply.sh [hunk 4]
```diff
@@ -444,6 +444,17 @@ run_kubectl_server_side_apply_tests() {
 }
 __EOF__
 
+  # Ensure the API server has recognized and started serving the associated CR API
+  local tries=5
+  for i in $(seq 1 $tries); do
+      local output=$(kubectl "${kube_flags_with_token[@]}" api-resources --api-group 'mygroup.example.com' -oname)
```
Comment (liggitt) [inline test/cmd/apply.sh:450]:
```suggestion
      local output=$(kubectl "${kube_flags[@]}" api-resources --api-group mygroup.example.com -o name)
```

File: test/cmd/apply.sh (hunk 2 above)
Comment (liggitt) [reply test/cmd/apply.sh:455]: this is similar, used in other places, and is easier to read:
```suggestion
      sleep "${i}"
```

**Comment by liggitt** [comment]
/milestone v1.19

test-only change addressing a flake clouding CI signal

**Comment by knight42** [comment]
Force push to address shellcheck errors
```
Errors from shellcheck:

In ./test/cmd/apply.sh line 450:
      local output=$(kubectl "${kube_flags[@]:?}" api-resources --api-group mygroup.example.com -oname)
            ^----^ SC2155: Declare and assign separately to avoid masking return values.


In ./test/cmd/apply.sh line 455:
      sleep $i
            ^-- SC2086: Double quote to prevent globbing and word splitting.

Did you mean: 
      sleep "$i"
```

**Pull request merged**